pytorch_lightning/trainer/trainer.py

```diff
@@ -497,6 +497,36 @@ def fit(
         val_dataloaders: Optional[EVAL_DATALOADERS] = None,
         datamodule: Optional[LightningDataModule] = None,
         train_dataloader=None,  # TODO: remove with 1.6
+    ) -> None:
+        try:
+            self._fit_impl(model, train_dataloaders, val_dataloaders, datamodule, train_dataloader)
+        except KeyboardInterrupt:
+            rank_zero_warn("Detected KeyboardInterrupt, attempting graceful shutdown...")
+            # user could press Ctrl+c many times... only shutdown once
+            if not self.interrupted:
+                self.state.status = TrainerStatus.INTERRUPTED
+                self.on_keyboard_interrupt()
+                # same treatment as below
+                self.accelerator.teardown()
+        except BaseException:
+            self.state.status = TrainerStatus.INTERRUPTED
+            if distributed_available() and self.world_size > 1:
+                # try syncing remaing processes, kill otherwise
+                self.training_type_plugin.reconciliate_processes(traceback.format_exc())
+            # give accelerators a chance to finish
+            self.accelerator.teardown()
+            self._on_exception()
+            # reset bookkeeping
+            self.state.stage = None
+            raise
```

Comment:
Do we see this exception handling behaviour diverging between functions?
Have you thought about using a common implementation?

Reply:
good point, will add